### ISSUE TITLE
Unable to start controller docker container using sosberlin/js7:controller-2-7-2

### ISSUE THREAD

**Issue by jon-anslow** [sos-berlin/js7#16]
Hi,

I have a local docker testing setup when I was previously using 2.7.0. This was all working fine but I decided to upgrade to 2.7.2. The JOC and agent containers started fine and I was able to get to the JOC console but my controller container is not starting. Below is how I am currently running it:

```bash
docker run \
      --hostname=js7-controller-primary \
      --network=js7 \
      --publish=15444:4444 \
      --env="RUN_JS_ID=jobscheduler" \
      --env="RUN_JS_JAVA_OPTIONS=-Xmx256m" \
      --name js7-controller-primary \
      sosberlin/js7:controller-2-7-2
```

My only change was to the image reference. It was sosberlin/js7:controller-2-7-0 but I have updated to sosberlin/js7:controller-2-7-2

Are you able to advise what the issue is and if I need to pass any additional configuration to the container to move to 2.7.2?

This is the output from:
**watchdog.log:**
```bash
"/usr/bin/java" -Xmx500m -Dfile.encoding=UTF-8 -classpath "/opt/sos-berlin.com/js7/controller/lib:/var/sos-berlin.com/js7/controller/config/patches/*:/var/sos-berlin.com/js7/controller/config/lib/*:/opt/sos-berlin.com/js7/controller/lib/patches/*:/opt/sos-berlin.com/js7/controller/lib/user_lib/*:/opt/sos-berlin.com/js7/controller/lib/sos/*:/opt/sos-berlin.com/js7/controller/lib/3rd-party/*" js7.controller.ControllerMain --id=jobscheduler --http-port="4444"  --config-directory="/var/sos-berlin.com/js7/controller/config" --data-directory="/var/sos-berlin.com/js7/controller"
...JS7 Controller(jobscheduler) is started with pid=136!
2024-11-19 11:58:31.506Z JS7 Controller 2.7.2
...JS7 Controller terminates with exit code 1
```

**controller.log:**
```bash
2024-11-19T11:58:30,695 Begin JS7 Controller · 2.7.2 · Java 17.0.13 · OpenJDK 64-Bit Server VM 17.0.13+11-alpine-r0 (500MiB) · Linux (Alpine Linux v3.20) · 06/9e (2 threads) 5.7GiB · pid=136 host=e6a95fa5f210
2024-11-19T11:58:30,824 info  js7.base.log.Logger - Controller 2.7.2 · Java 17.0.13 · OpenJDK 64-Bit Server VM 17.0.13+11-alpine-r0 (500MiB) · Linux (Alpine Linux v3.20) · 06/9e (2 threads) 5.7GiB · pid=136 host=e6a95fa5f210
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
2024-11-19T11:58:31,594 info  js7.base.configutils.Configs - Reading configuration file /var/sos-berlin.com/js7/controller/config/controller.conf
2024-11-19T11:58:31,789 info  js7.common.system.startup.ServiceMain - config=/var/sos-berlin.com/js7/controller/config data=/var/sos-berlin.com/js7/controller
2024-11-19T11:58:32,945 info  js7.base.log.Log4j - Shutdown at 2024-11-19 11:58:32.940704 (started at 2024-11-19 11:58:29Z 3.1s ago)
┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
```

**Comment by sosap** [comment]
Hi,

thank you. This doesn't provide a clear picture of the root cause as we find no error message from the logs. In the controller.log file we expect log output related to reading the Controller's journal files in its <data>/state directory. The fact that such output is missing might hint at a problem with volumes mounted to hold the journal files.

What output is reported from "docker logs" ? If you use the "docker run -it" command interactively, you should receive output with the console that might bettere explain what happened.

**Comment by sosap** [comment]
Hi,
a very bad situation. You skipped the instruction from the [https://kb.sos-berlin.com/display/JS7/JS7+-+Controller+Installation+for+Containers](https://kb.sos-berlin.com/display/JS7/JS7+-+Controller+Installation+for+Containers) article to persist the Controller's transaction data to a volume. 

As a result when restarting the container (independently from switching image/release) you lose all transactional data which includes orders that have proceeded to some point in a workflow.

You will have to redeploy workflows and related objects from the inventory. And you should cancel orders from the Daily Plan and resubmit.

The same applies to the Agent, without persistence of journal files it loses it's memory of orders.

Please assign a volume to Controller and Agent as indicated from

* [https://kb.sos-berlin.com/display/JS7/JS7+-+Controller+Installation+for+Containers](https://kb.sos-berlin.com/display/JS7/JS7+-+Controller+Installation+for+Containers)
* [https://kb.sos-berlin.com/display/JS7/JS7+-+Agent+Installation+for+Containers](https://kb.sos-berlin.com/display/JS7/JS7+-+Agent+Installation+for+Containers)

**Comment by jon-anslow** [comment]
Hi,

This is a short lived local environment purely for learning purposes but appreciate the requirement to add volumes to the controller and agent for persistence between container restarts.

However, I am still facing the same problem for the controller as outlined above with a volume attached.
**Note:** Testing the below command but with 2.7.0 works fine.

```bash
docker run -it \
      --hostname=js7-controller-primary \
      --network=js7 \
      --publish=15444:4444 \
      --env="RUN_JS_ID=jobscheduler" \
      --env="RUN_JS_JAVA_OPTIONS=-Xmx256m" \
      --mount="type=volume,src=js7-controller-primary-var,dst=/var/sos-berlin.com/js7/controller" \
      --name js7-controller-primary \
      sosberlin/js7:controller-2-7-2
```

Upon further investigation I believe the issue is that the private.conf file is missing and is not being copied from private.conf-example. 

```bash
> tree
.
└── var
    ├── config
    │   ├── controller.conf
    │   ├── controller.conf-example
    │   ├── lib
    │   ├── license
    │   ├── log4j2.xml-example
    │   ├── log4j2.xml-nofile
    │   ├── patches
    │   └── private
    │       ├── private.conf-example
    │       ├── trusted-pgp-keys
    │       │   └── sos.public-pgp-key.asc
    │       └── trusted-x509-keys
    │           └── sos.intermediate-ca.pem
    ├── logs
    │   ├── controller-debug.log
    │   ├── controller.log
    │   └── watchdog.log
    ├── state
    │   └── lock
    └── work

12 directories, 11 files
```

I have confirmed this by checking for the file within my volume and it has not been created, even though I see that controller .sh has an entry to copy the file from the example:

```bash
copy_file "${JS7_CONTROLLER_HOME}/var/config/private/private.conf-example" "${JS7_CONTROLLER_CONFIG_DIR}/private/private.conf"
```

I manually copied private.conf-example to become private.conf and now my container stays up and running.

Can you please check my findings? Do I need to set the JS7_CONTROLLER_HOME variable as part of the docker command?

**Comment by sosap** [comment]
Hi, 
thank you for the profound analysis. We confirm the issue originating from the installation script used for Controller containers:

* [SET-232: Controller installation script should provide private.conf file from example](https://change.sos-berlin.com/browse/SET-232)

The installation script does not ship with a specific release, but is available for download per release. 

* The fix has been applied to available downloads of the script.
* The fix has been applied to current deliveries 2.7.2 of Controller images.

**Comment by jon-anslow** [comment]
Hi,

Thank you for a fast turn around for this fix. I can confirm the 2.7.2 controller container is now working as expected. Many thanks for your help with the resolution of this.
